SSI Duplicate (#618)
OWASP/wstg: document/4-Web_Application_Security_Testing/07-Input_Validation_Testing/08-Testing_for_SSI_Injection.md

Server-Side Includes are directives that the web server parses before serving the page to the user. They represent an alternative to writing CGI programs or embedding code using server-side scripting languages, when there's only need to perform very simple tasks. Common SSI implementations provide directives(commands) to include external files, to set and print web server CGI environment variables, and to execute external CGI scripts or system commands.

When testing for SSI we are injecting SSI directives as user input and if SSI are enabled and there is no user input validation the server will execute the directive. This is very similar to a classical scripting language injection vulnerability, it occurs when user input is not properly validated and sanitized.

SSI can lead to a Remote Command Execution (RCE), however most webservers have the `exec` directive disabled by default.

> ### Testing examples

Reviewer: Can we remove this section from this part and use what's mainly needed in the How to Test section?

`<!--#exec cmd="ls" -->`
`<!--#exec cmd="OS_COMMAND" -->`

to include the output of a system command.
Returns the output of the supplied command.

Then, if the web server's SSI support is enabled, the server will parse these directives. In the default configuration, usually, most web servers don't allow the use of the `exec` directive to execute system commands.

As in every bad input validation situation, problems arise when the user of a web application is allowed to provide data that makes the application or the web server behave in an unforeseen manner. With regard to SSI injection, the attacker could provide input that, if inserted by the application (or maybe directly by the server) into a dynamically generated page, would be parsed as one or more SSI directives.

This is a vulnerability very similar to a classical scripting language injection vulnerability. One mitigation is that the web server needs to be configured to allow SSI. On the other hand, SSI injection vulnerabilities are often simpler to exploit, since SSI directives are easy to understand and, at the same time, quite powerful, e.g., they can output the content of files and execute system commands.

> ## Test Objectives

Reviewer: This is needed. Please set this back in as well :)

Author: This is now moved to the introduction before the how to test

Reviewer: This is needed as a section before the ## How to Test. It's the format we parse our checklist against.

Author: ## Test Objectives section returned

> There are more directives and variables which you can find [here](http://www.cheat-sheets.org/sites/ssi.su/)

Reviewer: Refer to this instead: https://httpd.apache.org/docs/current/howto/ssi.html

When testing for SSI we are injecting SSI directives as user input and if SSI are enabled and there is no user input validation the server will execute the directive. This is very similar to a classical scripting language injection vulnerability, it occurs when user input is not properly validated and sanitized.

The first thing to do when testing is finding if the web server actually supports SSI directives. Often, the answer is yes, as SSI support is quite common. To find out we just need to discover which kind of web server is running on our target, using classic information gathering techniques. If we have access to the code we can check if SSI directives are used in it or grep trough the webserver configuration files for specific keywords.

Another way of verifying that SSI are enabled, is checking for pages with the `.shtml` extension, since it is associated with SSI directives. Unfortunately, the use of the `.shtml` extension is not mandatory, so not having found any `.shtml` files doesn't necessarily mean that the target is not prone to SSI injection attacks.

The next step is determining all the possible user input vectors and testing if the SSI Injection is exploitable.

> To test this we have to find all the pages where user input is allowed. Possible input vectors may also include Headers and Cookies. Determine how the input is stored and used, i.e if the input is returned as an error message or page element and if it was modified in some way. If we have access to the source code we can easily determine where the input vectors are and how input is handled.

Reviewer 2: ```suggestion
First find all the pages where user input is allowed. Possible input vectors may also include headers and cookies. Determine how the input is stored and used, i.e if the input is returned as an error message or page element and if it was modified in some way. Access to the source code can help you to more easily determine where the input vectors are and how input is handled.
```

> Once we have a list of potential injection points, we can check if the input is correctly validated. We need to make sure that we can inject characters used in SSI directives:

Reviewer 2: ```suggestion
Once you have a list of potential injection points, you may determine if the input is correctly validated. Ensure it is possible to inject characters used in SSI directives:
```

`< ! # = / . " - > and [a-zA-Z0-9]`

> Testing if the SSI Injection is exploitable is very simillar to testing for XSS using

Reviewer 2: ```suggestion
Testing to see if SSI injection is exploitable is very similar to testing for XSS. You may use a string like the following:
```

`<script>alert("XSS")</script>`

> For SSI Injection we can use the directives listed below.

Reviewer 2: ```suggestion
For SSI injection we can use the directives listed below.
```

`<!--#echo var="VAR" -->`

> Returns the value of the variable. [List of SSI include variables](https://httpd.apache.org/docs/current/howto/ssi.html)

Reviewer 2: ```suggestion
This example returns the value of the variable. [See list of SSI include variables](https://httpd.apache.org/docs/current/howto/ssi.html).
```

`<!--#include virtual="FILENAME" -->`

> If the supplied file is a CGI script the directive will include the output of the CGI script. But the directive may also be used to include the content of a file or list files in a directory.

Reviewer 2: ```suggestion
If the supplied file is a CGI script, this directive will include the output of the CGI script. This directive may also be used to include the content of a file or list files in a directory.
```

`<!--#exec cmd="OS_COMMAND" -->`

Returns the output of the supplied command.

> There are more directives and variables which you can find [here](https://httpd.apache.org/docs/current/howto/ssi.html)

Reviewer 2: ```suggestion
You can find more directives and variables [here](https://httpd.apache.org/docs/current/howto/ssi.html).
```

Another way of verifying that SSI directives are enabled is by checking for pages with the `.shtml` extension, which is associated with SSI directives. The use of the `.shtml` extension is not mandatory, so not having found any `.shtml` files doesn't necessarily mean that the target is not vulnerable to SSI injection attacks.

The next step is determining all the possible user input vectors and testing to see if the SSI injection is exploitable.

First find all the pages where user input is allowed. Possible input vectors may also include headers and cookies. Determine how the input is stored and used, i.e if the input is returned as an error message or page element and if it was modified in some way. Access to the source code can help you to more easily determine where the input vectors are and how input is handled.

Once you have a list of potential injection points, you may determine if the input is correctly validated. Ensure it is possible to inject characters used in SSI directives such as `<!#=/."->` and `[a-zA-Z0-9]`

> The below example returns the value of the variable. Check the [references](#references) section to better assess the system at hand with every server's documentation.

Reviewer 2: ```suggestion
The below example returns the value of the variable. The [references](#references) section has helpful links with server-specific documentation to help you better assess a particular system.
```
